### ISSUE TITLE
Dynamic delay (delay extracted from return value/exception)

### ISSUE THREAD

**Issue by runeksvendsen** [Soostone/retry#63]
My app consumes an API which, for certain failure responses, will populate the [`Retry-After`](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Retry-After) header of the response with the number of seconds to wait before performing the next request.

As far as I can see, the interface to this library doesn't support adjusting the delay time based on information in the return value of (or the exception thrown by) the action that's retried. It seems that all the functions that apply a `RetryPolicy` statically deduce the waiting time based on information inside `RetryStatus`.

Would it be within the scope of this library to have e.g. `Control.Retry.retrying` support dynamically adjusting the delay time based on information in the return value of the action to run (the `b` type variable)?

Seems to me like the way to support this feature would be to change the `Bool` return type, that's used to indicate whether a request should be retried, to something along the lines of:

```haskell
data RetryAction
   = NoRetry                  -- ^ Don't retry
   | Retry                    -- ^ Retry with a delay calculated by the retry policy
   | RetryOverridingDelay Int -- ^ Retry, overriding the delay of the retry policy with the specified number of microseconds
```

**Comment by runeksvendsen** [comment]
I would like to implement the above suggestion. @MichaelXavier would you merge this?

**Comment by MichaelXavier** [comment]
Hmm. Right now we have a monoidal newtype over `RetryStatus -> m (Maybe Int)`, which couldn't really know how long to wait unless perhaps `m` was a stateful monad which included `Retry-After`. So I think your intuition is right to try to do this in the "runner" function like `retrying`, perhaps calling it `dynamicDelay` or something. Couldn't you even rewrite `retrying` and possibly `recovering` in terms of this lower-level function? I think they would just never use the `RetryOverridingDelay` case. I think this is worth doing and I'd accept the PR.

**Comment by runeksvendsen** [comment]
Yes, the current versions of `retrying` and `recovering` could indeed be expressed in terms of the theoretical `retryingDynamic` and `recoveringDynamic` functions, respectively, since a `Bool` can easily be translated into a `RetryAction`:

```haskell
toRetryAction :: Bool -> RetryAction
toRetryAction False = NoRetry
toRetryAction True = Retry
```

However, as far as I can see we would need to create both `retryingDynamic` and `recoveringDynamic`, ie. one for retrying without exceptions and one for retrying with exceptions.

One option is to make this a breaking change, so that both functions expect functions that return a `RetryAction` instead of a `Bool` (in the `Handler m` of `recovering` and the `chk` function passed to `retrying`). In addition we could export the above `toRetryAction` function, in order to make migration to this new version easier (library users would simply apply `toRetryAction` to the return values of their existing `chk` functions/`Handler`s). In this case I would also add an "upgrade guide" to the changelog, explaining this.

Another option would be to:
1. Keep the existing `retrying` and `recovering` functions
2. Add `retryingDynamic` and `recoveringDynamic`
3. Redefine `retrying` and `recovering` in terms of `retryingDynamic` and `recoveringDynamic`, respectively

**Comment by runeksvendsen** [comment]
I would also like to note that custom retry delays in the `Retry-After` HTTP header isn't the only place where overriding the result of the `RetryPolicy` is relevant. If, for example, the error returned by an action is an error internal to the client (e.g. [`InternalException`](https://www.stackage.org/haddock/lts-14.0/http-client-0.6.4/Network-HTTP-Client.html#v:InternalException)) there is no reason to do e.g. exponential backoff, and the client might as well retry immediately or after only a short delay.

**Comment by MichaelXavier** [comment]
I'd like to do your second option. I'm not entirely opposed to breaking changes but I don't think we should hassle users with having to fix all their use sites unless it really pays off to do so.

**Comment by runeksvendsen** [comment]
Great! I agree that it's probably best to minimize the amount of breaking changes.

I will open a PR when the code is done.

**Comment by runeksvendsen** [comment]
If you think of some good test cases please let me know.

**Comment by MichaelXavier** [comment]
Maybe have a canned sequence of delays, use the new feature with an action that pulls the delay out of the `RetryStatus` and asserts that the delays were the same and in the same order. Your monad could be just a pure `Writer` or `State` I think.

**Comment by runeksvendsen** [comment]
PR is up: #65

**Comment by MichaelXavier** [comment]
This was released in retry 0.8.1.0. Thanks again!
